File: csharp/1673.find-the-most-competitive-subsequence.cs
/*
 * @lc app=leetcode id=1673 lang=csharp
 *
 * [1673] Find the Most Competitive Subsequence
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int[] MostCompetitive(int[] nums, int k) {
        int n = nums.Length;
        var stack = new Stack<int>();
        for (int i = 0; i < n; ++i) {
            int curr = nums[i];
            while (stack.TryPeek(out var p) && p > curr && n - i + stack.Count > k) {
                stack.Pop();
            }
            if (stack.Count < k) {
                stack.Push(curr);
            }
        }
        return stack.Reverse().ToArray();
    }
}
// @lc code=end

File: java/1673.find-the-most-competitive-subsequence.java
/*
 * @lc app=leetcode id=1673 lang=java
 *
 * [1673] Find the Most Competitive Subsequence
 */

// @lc code=start
class Solution {
    public int[] mostCompetitive(int[] nums, int k) {
        int[] ret = new int[k];
        int index = -1, n = nums.length;
        for (int i = 0; i < n; ++i) {
            int x = nums[i];
            while (index >= 0 && ret[index] > x && n - i + index + 1 > k) {
                --index;
            }
            if (index < k - 1) {
                ret[++index] =  x;
            }
        }
        return ret;
    }
}
// @lc code=end

